mlx/compile.cpp

```diff
@@ -744,10 +740,15 @@ std::vector<array> compile_replace(
     trace_to_real.insert({trace_inputs[i].id(), inputs[i]});
   }
 
+  auto is_load = [](const Primitive& p) { return typeid(p) == typeid(Load); };
+
   for (auto& a : tape) {
-    // Arrays in the tape without primitives are constants
-    // and can be used directly
-    if (!a.has_primitive()) {
+    // Arrays in the tape without primitives are either:
+    // - inputs, which are already in the map
+    // - constants, which can be used directly
+    // - a load primitive which has no inputs and will become a constant
+    //   after the first eval
+    if (!a.has_primitive() || is_load(a.primitive())) {
```

Comment:
This change is worth commenting on:
- Previously if you loaded arrays from a file inside a compiled function then every call to the function would reload from the file.
- Now only the first call to the function loads and after that the loaded arrays become constants in the tape

This seems better to me.. though perhaps that is debatable. It is also used by `import_function` which makes `Load` primitives for constants and so we get lazy loading even with `import_function` which is pretty nice.

Lmk thoughts.. I can switch it so compile doesn't force load (more flexible but more dangerous).

Reply:
I was a bit torn but the more I think about it the more I like it. If someone wants to control the load, they can always pass the state as inputs.

Managing memory is also as easy as before. All you have to do is delete the function after it is called but before you eval (there is the small overhead of deserializing the function when loading again).

Reply:
It got me thinking about an optimization we can do in compile in general which is to eval branches of the graph that have no inputs on the leaves. It would be a trade-off for memory and compute so one would need to be careful there.. but there are some legitimate cases that it's come up for me. 

For example in some of our RoPE implementations we precompute the `self._freqs` but they aren't part of the `model.parameters()`. Naively compiling would recompute those at each call of the function. So you have to make sure you eval them before running the compiled function.

Reply:
Hmm interesting. It would be fairly easy to write as a part of simplify or a similar operation but how would we provide that functionality? Always doing it doesn't seem like the best option as it might require a lot of memory which the user explicitly doesn't want to keep around. Same goes for the `Load` indeed.

Reply:
One way of deciding is doing it based on user expectations. I'm not sure it's technically feasible. But there are two cases that we currently treat the same that I think people have quite different expectations about. Use load as an example:

```python
def fun1(x):
  return x + mx.load("y.safetensors")["y"]

y = mx.load("y.safetensors")["y"]

def fun2(x):
  return x + y
```

In `fun1` I expect the load to happen every time I call the function (as it does in eager mode). In `fun2` I expect the load to happen only once. 

For compile the behavior used to be like `fun1` and I switched it to `fun2` just because it's really unusual to write something like `fun1`.

But in theory we could try and distinguish the two cases. I think the expecatios are the same for any computation:

```python
def fun1(x):
  return x + complex_fun()

y = complex_fun()

def fun2(x):
  return x + y
```

Reply:
Getting that behavior in Python is probably doable because we can figure out which inputs are enclosed.. getting it in C++ is probably a lot harder / maybe not possible.